### ISSUE TITLE
Decoding an array using CodableCBORDecoder()

### ISSUE THREAD

**Issue by sijohans** [unrelentingtech/SwiftCBOR#56]
Hello, i'm trying to use the Decodable protocol to decode a cbor struct. The struct contain an optional array of integers. Just learning this (and swift) so i just have a very simple struct:
```swift
    public struct SimpleStruct: Codable {
        public let stringVal: String
        public let bytesVal: Data
        public let integer: Int
        public let booleanVal: Bool
        public let doubleVal: Double
        public let arrayVal: [Int]?
    }
```

I've written some "sanity" tests where i use the Encodable protocol to encode it, then using the Decodable protocol to decode that data. This works fine when there is no **arrayVal**.

See this code:
```swift
    func testSimpleStruct() {
        
        // Sanity 1
        let _ = try! CodableCBORDecoder().decode(SimpleStruct.self, from: CodableCBOREncoder().encode(SimpleStruct(stringVal: "foo", bytesVal: Data(hex: "aabbcc")!, integer: 4711, booleanVal: true, doubleVal: 3.14, arrayVal: nil)))

        // Sanity 2
        let _ = try! CodableCBORDecoder().decode(SimpleStruct.self, from: CodableCBOREncoder().encode(SimpleStruct(stringVal: "foo", bytesVal: Data(hex: "aabbcc")!, integer: 4711, booleanVal: true, doubleVal: 3.14, arrayVal: [])))
        
        // Sanity 3
        let _ = try! CodableCBORDecoder().decode(SimpleStruct.self, from: CodableCBOREncoder().encode(SimpleStruct(stringVal: "foo", bytesVal: Data(hex: "aabbcc")!, integer: 4711, booleanVal: true, doubleVal: 3.14, arrayVal: [1,2,3,-4711])))
    }
```

Both the sanity 1 and 2 test cases causes an exception.
```
Test Suite 'CborCoderTests' started at 2020-05-20 01:17:15.846
Test Case '-[CborCoderTests testSimpleStruct]' started.
Fatal error: 'try!' expression unexpectedly raised an error: Swift.DecodingError.typeMismatch(Swift.Optional<Any>, Swift.DecodingError.Context(codingPath: [], debugDescription: "cannot decode nil for key: CodingKeys(stringValue: \"arrayVal\", intValue: nil)", underlyingError: nil)): file CborCoderTests.swift, line 54
2020-05-20 01:17:27.735169+0200 xctest[48020:1831561] Fatal error: 'try!' expression unexpectedly raised an error: Swift.DecodingError.typeMismatch(Swift.Optional<Any>, Swift.DecodingError.Context(codingPath: [], debugDescription: "cannot decode nil for key: CodingKeys(stringValue: \"arrayVal\", intValue: nil)", underlyingError: nil)): file CborCoderTests.swift, line 54
Program ended with exit code: 9
```

I don't know whether this is a bug or if I simply misuse the codable protocol. Any suggestions?

Complete file if you want to reproduce:
```swift
import XCTest
import SwiftCBOR

extension Data {
    init?(hex: String) {
        let len = hex.count / 2
        var data = Data(capacity: len)
        for i in 0..<len {
            let j = hex.index(hex.startIndex, offsetBy: i*2)
            let k = hex.index(j, offsetBy: 2)
            let bytes = hex[j..<k]
            if var num = UInt8(bytes, radix: 16) {
                data.append(&num, count: 1)
            } else {
                return nil
            }
        }
        self = data
    }
}

extension Data {
    var hex: String {
        return reduce("") {$0 + String(format: "%02x", $1)}
    }
}

class CborCoderTests: XCTestCase {
    
    public struct SimpleStruct: Codable {
        public let stringVal: String
        public let bytesVal: Data
        public let integer: Int
        public let booleanVal: Bool
        public let doubleVal: Double
        public let arrayVal: [Int]?
    }
    
    func testSimpleStruct() {
        
        // Sanity 1
        let _ = try! CodableCBORDecoder().decode(SimpleStruct.self, from: CodableCBOREncoder().encode(SimpleStruct(stringVal: "foo", bytesVal: Data(hex: "aabbcc")!, integer: 4711, booleanVal: true, doubleVal: 3.14, arrayVal: nil)))

        // Sanity 2 - fail
        let _ = try! CodableCBORDecoder().decode(SimpleStruct.self, from: CodableCBOREncoder().encode(SimpleStruct(stringVal: "foo", bytesVal: Data(hex: "aabbcc")!, integer: 4711, booleanVal: true, doubleVal: 3.14, arrayVal: [])))
        
        // Sanity 3 - fail
        let _ = try! CodableCBORDecoder().decode(SimpleStruct.self, from: CodableCBOREncoder().encode(SimpleStruct(stringVal: "foo", bytesVal: Data(hex: "aabbcc")!, integer: 4711, booleanVal: true, doubleVal: 3.14, arrayVal: [1,2,3,-4711])))
        
        let tmp = SimpleStruct(stringVal: "foo", bytesVal: Data(hex: "aabbcc")!, integer: 4711, booleanVal: true, doubleVal: 3.14, arrayVal: [1,2,-3,4711])
        print(try! CodableCBOREncoder().encode(tmp).hex)
        
        // Definite map, no arrayVal
        let _ = try! CodableCBORDecoder().decode(SimpleStruct.self, from: Data(hex: "a56a626f6f6c65616e56616cf569737472696e6756616c63666f6f69646f75626c6556616cfb40091eb851eb851f67696e746567657219126768627974657356616c43aabbcc")!)
        
        // Using indefinite map, no arrayVal
        let _ = try! CodableCBORDecoder().decode(SimpleStruct.self, from: Data(hex: "bf6a626f6f6c65616e56616cf569737472696e6756616c63666f6f69646f75626c6556616cfb40091eb851eb851f67696e746567657219126768627974657356616c43aabbccff")!)
        
        // This works fine
        let _ = try! CBOR.decode(Array(Data(hex: "a669737472696e6756616c63666f6f6a626f6f6c65616e56616cf567696e746567657219126769646f75626c6556616cfb40091eb851eb851f68627974657356616c43aabbcc68617272617956616c84010222191267")!))
        
        // not this
        let _ = try! CodableCBORDecoder().decode(SimpleStruct.self, from: Data(hex: "a669737472696e6756616c63666f6f6a626f6f6c65616e56616cf567696e746567657219126769646f75626c6556616cfb40091eb851eb851f68627974657356616c43aabbcc68617272617956616c84010222191267")!)
        
    }
    
}
```

**Comment by lucaslenz** [comment]
I am having a very similar issue with the array type. It does not validate correctly. 

    struct SampleClass: Codable{
        private var id: [UInt8]
        private var type: String
    
        init(id: [UInt8], type: String){
            self.id = id
            self.type = type
        }
    
        init(from decoder: Decoder) throws {
            let container = try decoder.container(keyedBy: CodingKeys.self)
            let idData = try container.decode(Data.self, forKey: .id)
            self.id = [UInt8](idData)
            self.type = try container.decode(String.self, forKey: .type)
        }
    
        func getId() -> [UInt8]?{
            return id
        }
    
        mutating func setId(id: [UInt8]){
            self.id = id
        }
    
        func getType() -> String?{
            return type
        }
    
        mutating func setType(type: String){
            self.type = type
        }
    
        func serialize() -> [UInt8]{
            //
        }
    }
   
`let list = try container.decode([PublicKeyCredentialDescriptor].self, forKey: .excludeList)`

I also experimented with the `decodeIfPresent` method. `decodeIfPresent` returns nil although the key is available. If I try the `decode` method it also returns nil. Passing a single value of that type is working as the SampleClass init from decoder method is called. On the array version this method is not called and thus the values are not extracted.

**Comment by ArturFriesen** [comment]
Need to push this. There still seems to be a problem with optional nested types (e.g. optional array or optional non primitive types). I'm getting the error "cannot decode nil for key: CodingKeys(stringValue: \"foo\", intValue: nil)"

**Comment by stoprocent** [comment]
bump

**Comment by hamchapman** [comment]
@stoprocent have you tried using the library recently? It will probably work for you. I believe the original problem reported in this issue is all fixed now

**Comment by hamchapman** [comment]
I double checked and

**Comment by hamchapman** [comment]
This was fixed in #80
